externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td

```diff
@@ -0,0 +1,65 @@
+#ifndef TORCH_MLIR_DIALECT_TCP_OPS
+#define TORCH_MLIR_DIALECT_TCP_OPS
+
+include "torch-mlir-dialects/Dialect/Tcp/IR/TcpOpsBase.td"
+
+include "mlir/IR/OpBase.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+def Tcp_TanhOp : Tcp_UnaryOp<"tanh", [NoSideEffect]> {
+  let summary = "Computes tanh of input, elementwise";
+
+  let description = [{
+    Elementwise unary op to compute tanh.
+  }];
+
+  let arguments = (ins
+    Tcp_Tensor:$in
+  );
+
+  let results = (outs
+    Tcp_Tensor:$out
+  );
+
+  let assemblyFormat = "$in attr-dict `:` type($in) `->` type($out)";
+}
+
+def Tcp_AddOp : Tcp_BinaryOp<"add", [NoSideEffect]> {
```

Comment:
shape inference fns, infer type op interface.

cannot be marked NoSideEffect since it has UB (or do we want a separate Speculatable trait? IIRC this was broken in MLIR core). Sanjoy is the expert on this stuff.

Reply:
Yes, this cannot be `NoSideEffect` so I'd remove it from now.  We'll eventually have to figure out a more nuanced way to express this trait.

Reply:
Good point. Removed.